### PR TITLE
Segment Status resolved

### LINKED ISSUE #1456: Status changes to "Unused" when the Segment details page is refreshed
**Version where bug was found:**
5.1.5

**Describe the bug**
The status changes to "Unused" when the Segment details page is refreshed.

**To Reproduce**
1. Go to the Segment details page of a segment with either "Used" or "Global" status.
2. Refresh the page.
3. Observe that the status has changed to "Unused".

**Expected behavior**
The status should remain unchanged after the page is refreshed.

**Screenshots**
Screenshot 2024-04-16 at 4 12 37 PM

### PR DESCRIPTION
**Pull request by JayMehta11**
This PR resolves #1456. 

**Issues resolved:**

-  The 'globalStatus' page retains its status and will not change to 'Unused' when refreshed.
-  If segments used in an experiment have subsegments, their statuses will also be updated recursively through all levels.
-  If segments are used in a global segment, their statuses will be changed to 'Used'. Similarly, if these segments have subsegments, their statuses will also be updated accordingly.
- Created a new, optimised query for fetching statuses on the segments page and individual segment page.

### CODE REVIEW

File: backend/packages/Upgrade/src/api/services/SegmentService.ts [hunk 1]
```diff
@@ -84,7 +84,37 @@ export class SegmentService {
       .andWhere({ id })
       .getOne();
 
-    return segmentDoc;
+    if (!segmentDoc) {
+      return undefined;
+    }
+
+    const segmentsUsedList = [];
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/services/SegmentService.ts:91]: This logic is checking direct sub-segment which is not an entirely correct solution. 
@ppratikcr7 can you help Jay with this

File: backend/packages/Upgrade/test/unit/services/SegmentService.test.ts [hunk 2]
```diff
@@ -219,7 +219,7 @@ describe('Segment Service Testing', () => {
 
   it('should get segment by id', async () => {
     const segments = await service.getSegmentById(seg1.id, logger);
-    expect(segments).toEqual(seg1);
+    expect(segments).toEqual({ ...seg1, status: 'Unused' });
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/test/unit/services/SegmentService.test.ts:222]: Use the Enum

**Review by VivekFitkariwala — CHANGES_REQUESTED** [review]

File: backend/packages/Upgrade/src/api/services/SegmentService.ts [hunk 3]
```diff
@@ -73,6 +76,17 @@ export class SegmentService {
     return queryBuilder;
   }
 
+  public async getAllSegmentsAndSubsegments(logger: UpgradeLogger): Promise<Segment[]> {
+    logger.info({ message: `Find all segments` });
```
Comment (RidhamShah) [inline backend/packages/Upgrade/src/api/services/SegmentService.ts:80]: This logger should be different from getAllSegments

**Review by RidhamShah — CHANGES_REQUESTED** [review]

File: backend/packages/Upgrade/src/api/controllers/SegmentController.ts [hunk 4]
```diff
@@ -1,13 +1,16 @@
...
 import { Segment } from '../models/Segment';
-import { SERVER_ERROR } from 'upgrade_types';
+import { SEGMENT_STATUS, SERVER_ERROR } from 'upgrade_types';
 import { isUUID } from 'class-validator';
 import { AppRequest } from '../../types';
 import { SegmentFile, SegmentIds, SegmentImportError, SegmentInputValidator } from './validators/SegmentInputValidator';
 import { ExperimentSegmentInclusion } from '../models/ExperimentSegmentInclusion';
 import { ExperimentSegmentExclusion } from '../models/ExperimentSegmentExclusion';
 
+export class SegmentStatus extends Segment {
```
[the remark this answers is not in the document]
Comment (JayMehta11) [reply backend/packages/Upgrade/src/api/controllers/SegmentController.ts]: code updated

File: backend/packages/Upgrade/src/api/services/SegmentService.ts [hunk 5]
```diff
@@ -73,7 +73,7 @@ export class SegmentService {
     return queryBuilder;
   }
 
-  public async getSegmentById(id: string, logger: UpgradeLogger): Promise<Segment> {
+  public async getSegmentById(id: string, logger: UpgradeLogger): Promise<SegmentStatus | undefined> {
```
[the remark this answers is not in the document]
Comment (JayMehta11) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts]: code updated

File: backend/packages/Upgrade/src/api/services/SegmentService.ts (hunk 1 above)
Comment (JayMehta11) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts]: new query had been created and this code has been removed

File: backend/packages/Upgrade/test/unit/services/SegmentService.test.ts (hunk 2 above)
Comment (JayMehta11) [reply backend/packages/Upgrade/test/unit/services/SegmentService.test.ts]: code has been removed

File: backend/packages/Upgrade/src/api/services/SegmentService.ts (hunk 3 above)
Comment (JayMehta11) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts]: code updated

File: backend/packages/Upgrade/src/api/services/SegmentService.ts [hunk 6]
```diff
@@ -125,32 +143,45 @@ export class SegmentService {
...
+        const subSegmentIds = segmentMap.get(segmentId) || [];
+        subSegmentIds.forEach((subSegmentId) => collectSegmentIds(subSegmentId));
+      };
 
       if (allExperimentSegmentsInclusion) {
         allExperimentSegmentsInclusion.forEach((ele) => {
-          const subSegments = ele.segment.subSegments;
-          segmentsUsedList.push(...subSegments.map((subSegment) => subSegment.id));
+          collectSegmentIds(ele.segment.id);
+          ele.segment.subSegments.forEach((subSegment) => collectSegmentIds(subSegment.id));
```
Comment (RidhamShah) [inline backend/packages/Upgrade/src/api/services/SegmentService.ts:170]: We don't need this line of code. The above collectSegmentIds would have called its subSegments.

File: backend/packages/Upgrade/src/api/services/SegmentService.ts [hunk 7]
```diff
@@ -125,32 +143,45 @@ export class SegmentService {
...
         });
       }
 
       if (allExperimentSegmentsExclusion) {
         allExperimentSegmentsExclusion.forEach((ele) => {
-          const subSegments = ele.segment.subSegments;
-          segmentsUsedList.push(...subSegments.map((subSegment) => subSegment.id));
+          collectSegmentIds(ele.segment.id);
+          ele.segment.subSegments.forEach((subSegment) => collectSegmentIds(subSegment.id));
```
Comment (RidhamShah) [inline backend/packages/Upgrade/src/api/services/SegmentService.ts:177]: Same as above

File: backend/packages/Upgrade/src/api/services/SegmentService.ts (hunk 6 above)
Comment (JayMehta11) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts:170]: we need this line as ele.segment.id is a private segment.

File: backend/packages/Upgrade/src/api/services/SegmentService.ts (hunk 7 above)
Comment (JayMehta11) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts:177]: The answer is mentioned above

File: backend/packages/Upgrade/src/api/services/SegmentService.ts (hunk 1 above)
Comment (VivekFitkariwala) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts]: The private segments are also stored in the Segment table. The question is why are you filtering private segments in getAllSegmentsAndSubsegments?

Comment (JayMehta11) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts]: Private segments are called ahead in another function where Experiment's segments are called and we check it in nested loop, in that we get details from the subsegments.

File: backend/packages/Upgrade/src/api/services/SegmentService.ts [hunk 8]
```diff
@@ -73,6 +76,17 @@ export class SegmentService {
     return queryBuilder;
   }
 
+  public async getAllSegmentsAndSubsegments(logger: UpgradeLogger): Promise<Segment[]> {
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/services/SegmentService.ts:79]: Rename -> `getAllPublicSegmentsAndSubsegments`

File: backend/packages/Upgrade/src/api/services/SegmentService.ts [hunk 9]
```diff
@@ -108,13 +122,17 @@ export class SegmentService {
...
+    const allSegmentData = await this.getAllSegmentsAndSubsegments(logger);
     const segmentData = await this.getSegmentById(segmentId, logger);
-    return (await this.getSegmentStatus([segmentData])).segmentsData[0];
+    const segmentWithStatus = (await this.getSegmentStatus(allSegmentData)).segmentsData.find(
+      (segment: Segment) => segment.id === segmentId
+    );
+    return { ...segmentData, status: segmentWithStatus.status };
   }
 
   public async getAllSegmentWithStatus(logger: UpgradeLogger): Promise<getSegmentData> {
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/services/SegmentService.ts:134]: Now getSegmentStatus will always need all the segment data that should be part of this function as well.

File: backend/packages/Upgrade/src/api/services/SegmentService.ts [hunk 10]
```diff
@@ -125,32 +143,45 @@ export class SegmentService {
         this.getExperimentSegmentExclusionData(),
       ]);
 
-      const segmentsUsedList = [];
+      const segmentMap = new Map<string, string[]>();
+      segmentsData.forEach((segment) => {
+        if (segment.id === globalExcludeSegment.id) {
+          segmentMap.set(segment.id, []);
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/services/SegmentService.ts:149]: Can't the global segment have sub-segments?

**Review by VivekFitkariwala — CHANGES_REQUESTED** [review]
A question before I approve the PR

File: backend/packages/Upgrade/src/api/services/SegmentService.ts (hunk 10 above)
Comment (JayMehta11) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts:149]: Should we have global segment's sub-segments status as "Used"?

File: backend/packages/Upgrade/src/api/services/SegmentService.ts (hunk 9 above)
Comment (JayMehta11) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts:134]: Yes, we will need that.

File: backend/packages/Upgrade/src/api/services/SegmentService.ts (hunk 8 above)
Comment (JayMehta11) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts:79]: Updated

**Comment by danoswaltCL** [comment]
@VivekFitkariwala can we merge this now?

**Review by VivekFitkariwala — APPROVED** [review]

**Pull request merged**